Fix #4388: Split IRChecker into ClassDefChecker and IRChecker

gzm0 opened the pull request.

gzm0 commented:
First draft of how this could look like. Observations:

- Found #4409: Good we started this :)
- We should consider a `JSParamDef`.
- I'm toying with the idea of splitting `ClassDef#memberDefs` more, maybe even use an order preserving map type where applicable. WDYT?
- An even bolder idea would be to make different concrete subclasses of `ClassDef` based on kind (making many of the "you can't have x here checks moot).
- We get caching of IR pre link IR checking for free.
- I'm bothered that without tree traversal, we have to decide between assertions in trees or checking later.
- We should probably call this `ClassDef` checker.

sjrd commented:
> * We should consider a `JSParamDef`.
> * I'm toying with the idea of splitting `ClassDef#memberDefs` more, maybe even use an order preserving map type where applicable. WDYT?
> * An even bolder idea would be to make different concrete subclasses of `ClassDef` based on kind (making many of the "you can't have x here checks moot).

For all of these, I think the most important criteria is: would those changes ease the non-checker components (base linker, optimizer, emitter), or make them more difficult? IMO we shouldn't make the job of the "active" components harder just to make the IR checker(s) easier or the IR data structures more type-safe.

> * We get caching of IR pre link IR checking for free.
> * I'm bothered that without tree traversal, we have to decide between assertions in trees or checking later.

Actually the caching for free is an argument in favor of doing as much IR checking pre-link, so including a tree traversal.

> * We should probably call this `ClassDef` checker.

Yes :-p

gzm0 commented:
> Actually the caching for free is an argument in favor of doing as much IR checking pre-link, so including a tree traversal.

Yeah... Although, we wont get away without another traversal :( So if the execution time is traversal bound (and not what we actually do), checking later might be faster in practical scenarios. I do not expect `fullLinkJS` to run often incrementally... (but then, who knows :P).

> For all of these, I think the most important criteria is: would those changes ease the non-checker components (base linker, optimizer, emitter), or make them more difficult? IMO we shouldn't make the job of the "active" components harder just to make the IR checker(s) easier or the IR data structures more type-safe.

I'll prepare some PRs so we can discuss about concrete changes :)

gzm0 commented:
@sjrd WDYT about the changes I pointed out. Shall I split them into a separate commit?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -518,33 +507,26 @@ private final class Analyzer(config: CommonPhaseConfig,
+          val superCl = superClass.get // checked by ClassDef checker.
+          superCl.kind match {
+            case ClassKind.JSClass | ClassKind.NativeJSClass =>
+              superClass // ok
+
+            case _ =>
```
New spurious blank line here.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/BaseLinker.scala:
```diff
@@ -33,19 +33,19 @@ import Analysis._
  *  [[standard.LinkedClass LinkedClass]]es. Does a dead code elimination pass.
  */
-final class BaseLinker(config: CommonPhaseConfig) {
+final class BaseLinker(config: CommonPhaseConfig, checkIR: Boolean) {
```
It seems it wouldn't cost anything to add a deprecated constructor with the old signature here (defaulting to `checkIR = false`).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/IRLoader.scala:
```diff
@@ -16,20 +16,28 @@ import scala.collection.mutable
+  private var logger: Logger = _
   private val cache = mutable.Map.empty[ClassName, ClassDefAndInfoCache]
 
-  def update(irInput: Seq[IRFile])(implicit ec: ExecutionContext): Future[this.type] = {
```
Here we could also keep this signature as deprecated.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/IRLoader.scala:
```diff
@@ -121,6 +131,13 @@ private final class ClassDefAndInfoCache {
           version.get != newVersion.get) {
         version = newVersion
```
Should we move this assignment after the one to `cacheUpdate`, now that the latter can "likely" throw?

I don't think the `IRLoader` is protected by the "any exception invalidates the whole state" rule, is it?

gzm0 replied:
The assignment to `cacheUpdate` will not throw. The checking is asynchronous: So the `Future` representing the load will be in a failed state. IMO this is consistent: A new version of the file could be a non failed state (although I don't think the linker is able to use this).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/BaseLinker.scala:
```diff
@@ -33,19 +33,19 @@ import Analysis._
   def link(irInput: Seq[IRFile],
       moduleInitializers: Seq[ModuleInitializer], logger: Logger,
-      symbolRequirements: SymbolRequirement, checkIR: Boolean)(
+      symbolRequirements: SymbolRequirement)(
```
Could we preserve the old signature as deprecated in a meaningful way?

sjrd commented on linker/shared/src/test/scala/org/scalajs/linker/checker/ClassDefCheckerTest.scala:
```diff
@@ -0,0 +1,147 @@
+          )),
+        "duplicate constructor method '<init>(java.lang.String)void'")
+  }
+  }
```
There's something wrong here.

sjrd commented on linker/shared/src/test/scala/org/scalajs/linker/checker/ClassDefCheckerTest.scala:
```diff
@@ -0,0 +1,147 @@
+  @Test
+  def noDuplicateConstructors(): Unit = {
+    val BoxedStringType = ClassType(BoxedStringClass)
+    val stringCtorName = MethodName.constructor(List(ClassRef(BoxedStringClass)))
```
```suggestion
    val stringCtorName = MethodName.constructor(List(T))
```

sjrd reviewed (changes requested):
First pass on everything except the files `IRChecker.scala` into `ClassDefChecker.scala` (i.e., the core change).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+
+    def checkApplyGeneric(methodName: MethodName, args: List[Tree]): Unit = {
+      val paramRefs = methodName.paramTypeRefs.size
+      if (args.size != paramRefs)
+        reportError(i"Arity mismatch: $paramRefs expected but "+
+            i"${args.size} found")
```
```suggestion
      if (args.size != paramRefs)
        reportError(i"Arity mismatch: $paramRefs expected but ${args.size} found")
```
or
```suggestion
      if (args.size != paramRefs) {
        reportError(
            i"Arity mismatch: $paramRefs expected but ${args.size} found")
      }
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+    implicit val ctx = ErrorContext(tree)
+
+    def checkApplyGeneric(methodName: MethodName, args: List[Tree]): Unit = {
+      val paramRefs = methodName.paramTypeRefs.size
```
```suggestion
      val paramRefsCount = methodName.paramTypeRefs.size
```
?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+      case VarDef(ident, _, vtpe, mutable, rhs) =>
+        checkDeclareLocalVar(ident)
+        checkTree(rhs, env)
+        env.withLocal(LocalDef(ident.name, vtpe, mutable))
```
This is the only `case` that returns an env different from `env`. Could we put that separately from the main `tree match`?

gzm0 replied:
Done.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+            if (env.locals.get(name).exists(!_.mutable))
+              reportError(i"Assignment to immutable variable $name.")
+
+          case _: Select | _: JSPrivateSelect | _: SelectStatic |
```
```suggestion
          case _:Select | _:JSPrivateSelect | _:SelectStatic |
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+        env.locals.get(name).fold[Unit] {
+          reportError(i"Cannot find variable $name in scope")
+        } { localDef =>
+          if (tree.tpe != localDef.tpe)
+            reportError(i"Variable $name of type ${localDef.tpe} "+
+                i"typed as ${tree.tpe}")
```
Probably fine to write on one line:
```suggestion
          if (tree.tpe != localDef.tpe)
            reportError(i"Variable $name of type ${localDef.tpe} typed as ${tree.tpe}")
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+        }
+        env
+
+      case This() =>
+        // TODO: Force type to be exact? This would allow to remove thisType tracking from IRChecker.
```
Did you still intend to implement this? It seems like it would be desirable.

gzm0 replied:
Yes, but it's a backwards incompatible change. So I'd need to build the entire hack processing. I was thinking to just leave it like that. If you feel it is worth tracking explicitly, I can file a follow-up issue.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+        checkTree(array, env)
+        env
+
+      case ArraySelect(array, index) =>
+        if (!array.tpe.isInstanceOf[ArrayType])
+          reportError(i"Array type expected but ${array.tpe} found")
+        checkTree(index, env)
+        env
```
Looks like this is missing a
```scala
checkTree(array, env)
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+
+    // `name` / `originalName` do not need checking.
+
+    checkKind(classDef.kind)
```
I thought it'd be nice to be explicit about the fields we pass in. But fair enough, the name of the methods already makes it clear enough :) I have removed the params.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+      checkTrees(args, env)
+    }
+
+    tree match {
```
Done. This ordering is likely still from a time where the IRChecker distinguished between stat / expr trees.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala:
```diff
@@ -1219,27 +729,6 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
               "Is/AsInstanceOf")
         }
 
-      case NoType | NullType | NothingType | _:RecordType =>
-        reportError(i"$tpe is not a valid target type for Is/AsInstanceOf")
-
-      case tpe: ArrayType =>
-        checkArrayType(tpe)
-
-      case _ =>
-        // ok
```
Good point. I have improved the comment.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:
```diff
@@ -0,0 +1,964 @@
+    if (namespace.isPrivate)
+      reportError("A field cannot be private")
+    if (namespace.isConstructor)
+      reportError("A field cannot be a constuctor")
```
No. Done.

gzm0 commented:
Regarding deprecated methods: I'm not a fan of attempting to keep compatibility in a part of the codebase where we clearly do not guarantee this (also, it's not actually obvious how to do it).

With that, I believe I have addressed all the comments. I have not resolved the ones where the changes were non-trivial so it's easier to keep track of what to look at.

gzm0 commented:
Also: I expect there is going to be another CI failure. I didn't fix it, because the logs were gone already.

gzm0 commented:
I have removed the braces after the definition of `Apply` (this caused scalastyle to fail).

gzm0 commented:
CI is hit by #4672.
